### PR TITLE
[MBL-16905][Student] Offline File Details

### PR DESCRIPTION
**Pull request by domonkosadam**
Test plan: Steps are in the ticket

refs: MBL-16905
affects: Student
release note: none

## Checklist

- [x] Tested in light mode

### CODE REVIEW

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

File: apps/student/src/main/java/com/instructure/student/features/files/details/FileDetailsFragment.kt [hunk 1]
```diff
@@ -139,6 +143,17 @@ class FileDetailsFragment : ParentFragment() {
...
+        if (isOffline()) {
+            binding.downloadButton.visibility = View.GONE
+        }
+        else {
+            binding.downloadButton.visibility = View.VISIBLE
+        }
+    }
+
+    private fun isOffline(): Boolean {
+        return networkStateProvider.isOnline().not()
```
Comment (hermannakos) [inline apps/student/src/main/java/com/instructure/student/features/files/details/FileDetailsFragment.kt:156]: You don't have to inject the NetworkStateProvider for this. you can use the repository here to check if we're online.

File: apps/student/src/main/java/com/instructure/student/features/files/details/FileDetailsFragment.kt [hunk 2]
```diff
@@ -139,6 +143,17 @@ class FileDetailsFragment : ParentFragment() {
                 requestPermissions(PermissionUtils.makeArray(PermissionUtils.WRITE_EXTERNAL_STORAGE), PermissionUtils.WRITE_FILE_PERMISSION_REQUEST_CODE)
             }
         }
+
+        if (isOffline()) {
+            binding.downloadButton.visibility = View.GONE
```
Comment (hermannakos) [inline apps/student/src/main/java/com/instructure/student/features/files/details/FileDetailsFragment.kt:148]: You can use the View::setVisible extension here.

File: apps/student/src/main/java/com/instructure/student/features/files/details/FileDetailsLocalDataSource.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import com.instructure.canvasapi2.models.FileFolder
+import com.instructure.pandautils.room.offline.daos.FileFolderDao
+import com.instructure.pandautils.room.offline.daos.LocalFileDao
+import okhttp3.ResponseBody
+
+class FileDetailsLocalDataSource(
+    private val fileFolderDao: FileFolderDao,
+    private val localFileFolderDao: LocalFileDao,
+) : FileDetailsDataSource {
+    override suspend fun markAsRead(canvasContext: CanvasContext, moduleId: Long, itemId: Long, forceNetwork: Boolean): ResponseBody? {
```
Comment (hermannakos) [inline apps/student/src/main/java/com/instructure/student/features/files/details/FileDetailsLocalDataSource.kt:31]: You're always calling the networkDataSource for this call, so you this localDataSource doesn't need to have it at all to keep it clean. You could just remove it from the interface and just add it to the networkDataSource.

**Review by hermannakos — CHANGES_REQUESTED** [review]
Great job! Just some minor things to make it cleaner.

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

File: apps/student/src/main/java/com/instructure/student/features/files/details/FileDetailsFragment.kt [hunk 4]
```diff
@@ -139,6 +140,17 @@ class FileDetailsFragment : ParentFragment() {
                 requestPermissions(PermissionUtils.makeArray(PermissionUtils.WRITE_EXTERNAL_STORAGE), PermissionUtils.WRITE_FILE_PERMISSION_REQUEST_CODE)
             }
         }
+
+        if (isOffline()) {
+            binding.downloadButton.setVisible(false)
```
Comment (hermannakos) [inline apps/student/src/main/java/com/instructure/student/features/files/details/FileDetailsFragment.kt:145]: You could pass the condition to the setVisible instead. `binding.downloadButton.setVisible(!isOffline())` or even better, making not inverting the repository isOnline and using that instead like `binding.downloadButton.setVisible(repository.isOnline())`

**Review by hermannakos — CHANGES_REQUESTED** [review]

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Review by hermannakos — APPROVED** [review]

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

File: apps/student/src/main/java/com/instructure/student/features/modules/progression/ModuleProgressionRepository.kt [hunk 5]
```diff
@@ -66,4 +66,9 @@ class ModuleProgressionRepository(
         val courseSyncSettings = courseSyncSettingsDao.findById(courseId)
         return courseSyncSettings?.tabs?.filter { it.value }?.keys ?: emptySet()
     }
+
+    suspend fun getSyncedFileIds(courseId: Long): List<Long> {
```
Comment (tamaskozmer) [inline apps/student/src/main/java/com/instructure/student/features/modules/progression/ModuleProgressionRepository.kt:70]: You should write tests for this.

File: apps/student/src/main/java/com/instructure/student/features/modules/util/ModuleUtility.kt [hunk 6]
```diff
@@ -127,6 +144,21 @@ object ModuleUtility {
...
     }
 
+    private fun createFileDetailsFragmentWithOfflineCheck(
+        isOnline: Boolean,
+        course: Course,
+        item: ModuleItem,
+        syncedFiles: List<Long>,
+        context: Context,
+        creationBlock: () -> Fragment?
+    ): Fragment? {
```
Comment (tamaskozmer) [inline apps/student/src/main/java/com/instructure/student/features/modules/util/ModuleUtility.kt:154]: I think you can move the whole logic into this function. We only call this in one instance so the creation block  is always the same.

File: apps/student/src/main/java/com/instructure/student/features/modules/util/ModuleUtility.kt [hunk 7]
```diff
@@ -100,11 +101,27 @@ object ModuleUtility {
             }
         }
         "File" -> { // TODO Handle offline availability after files sync
-            val url = removeDomain(item.url)
-            if (moduleObject == null) {
-                FileDetailsFragment.newInstance(FileDetailsFragment.makeRoute(course, url!!, item.contentId))
-            } else {
-                FileDetailsFragment.newInstance(FileDetailsFragment.makeRoute(course, moduleObject, item.id, url!!, item.contentId))
+            createFileDetailsFragmentWithOfflineCheck(isOnline, course, item, syncedFileIds, context) {
```
Comment (tamaskozmer) [inline apps/student/src/main/java/com/instructure/student/features/modules/util/ModuleUtility.kt:104]: You can also add a test for this logic like for the other screens.

**Review by tamaskozmer — CHANGES_REQUESTED** [review]

**Review by kdeakinstructure — APPROVED** [review]
QA 👍  (with the conclusion of our meeting on Wednesday that we leave it as it is now, so the user have to select the exact module items, for example files as well, if the module item is a file) to be able to see it offline within the module.)

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Review by tamaskozmer — APPROVED** [review]

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Pull request merged**